### PR TITLE
http_client: implement NO_PROXY support

### PR DESCRIPTION
**Pull request by erain**
implement NO_PROXY support in fluent-bit:
- `NO_PROXY` is a standard environment variable for the URL should not go through the proxy, when `HTTP_PROXY` is used. 
- `NO_PROXY` is "A comma-separated list of host names that shouldn't go through any proxy is set in (only an asterisk, '*' matches all hosts)" ([reference](https://superuser.com/questions/944958/are-http-proxy-https-proxy-and-no-proxy-environment-variables-standard))

FIXES https://github.com/fluent/fluent-bit/issues/2805

----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:
- [N/A] Example configuration file for the change
- [N/A] Debug log output from testing the change

- [ ] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

**Documentation**

- [ ] Documentation required for this feature

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

### CODE REVIEW

**Comment by erain** [comment]
I successfully tested this feature in our environment, @pranavmarla could you help also test in your environment?

My testing pods are running in a kubernetes environment with environment variables something like below, with a squid http proxy. And please note there should not be spaces in the `NO_PROXY` string.

```
        - name: HTTP_PROXY
          value: http://abc:123@<proxy-host>:<proxy-port>
        - name: NO_PROXY
          value: 127.0.0.1,localhost,kubernetes.default.svc.cluster.local
```

**Comment by qingling128** [comment]
Hi @edsiper , could you help prioritize the code review of this PR? It is required for our GA as well.

**Comment by pranavmarla** [comment]
> I successfully tested this feature in our environment, @pranavmarla could you help also test in your environment?
> 
> My testing pods are running in a kubernetes environment with environment variables something like below, with a squid http proxy. And please note there should not be spaces in the `NO_PROXY` string.
> 
> ```
>         - name: HTTP_PROXY
>           value: http://abc:123@<proxy-host>:<proxy-port>
>         - name: NO_PROXY
>           value: 127.0.0.1,localhost,kubernetes.default.svc.cluster.local
> ```

@erain That's awesome, glad to hear it!

Unfortunately though, it might be some time till I'm in a position to test this out -- my EKS environment, Fluent Bit setup, etc. has long since been torn down, so I'd need to recreate all that.
Also, to test out your changes, I assume I'd need to build Fluent Bit from source (i.e. your branch) and then create a custom Docker image containing that custom Fluent Bit to reference in my Kubernetes deployment, all of which will also take time.
So basically, please don't wait on me to merge this change!

File: src/flb_upstream.c [hunk 1]
```diff
@@ -237,6 +237,39 @@ struct flb_upstream *flb_upstream_create(struct flb_config *config,
     return u;
 }
 
+/*
+ * Checks whehter a destinate URL should be proxied.
+ */
+int flb_should_proxy_for_host(const char *host, const char *proxy, const char *no_proxy)
```
Comment (edsiper) [inline src/flb_upstream.c:243]: the function name does not follow the interface pattern, please rename it to something like ```flb_upstream_needs_proxy()```

File: src/flb_upstream.c [hunk 2]
```diff
@@ -237,6 +237,39 @@ struct flb_upstream *flb_upstream_create(struct flb_config *config,
...
+        return FLB_TRUE;
+    }
+
+    /* NO_PROXY=`*`, it matches all hosts. */
+    if (strcmp(no_proxy, "*") == 0) {
+        return FLB_FALSE;
+    }
+
+    /* check the URL list in the NO_PROXY  */
+    char *no_proxy_url = strtok(no_proxy, ",");
```
Comment (edsiper) [inline src/flb_upstream.c:261]: - variable needs to be declared on top

- do not use strtok, instead go with flb_slist_split_tokens() API

File: include/fluent-bit/flb_str.h [hunk 3]
```diff
@@ -58,4 +58,13 @@ static inline char *flb_strndup(const char *s, size_t n)
     return str;
 }
 
+/* emptyval checks whether a string has a non-null value "". */
+static inline int flb_str_emptyval(const char *s)
```
Comment (edsiper) [inline include/fluent-bit/flb_str.h:62]: this function is only called from flb_config.c file, let's move it there in the .c file on top as a static inline.

**Review by edsiper — CHANGES_REQUESTED** [review]

File: include/fluent-bit/flb_str.h (hunk 3 above)
Comment (erain) [reply include/fluent-bit/flb_str.h]: DONE.

File: src/flb_upstream.c (hunk 2 above)
Comment (erain) [reply src/flb_upstream.c]: DONE. Thanks for letting me know, I used `flb_slist_split_string` for this now and also enhanced its unit tests.

File: src/flb_upstream.c (hunk 1 above)
Comment (erain) [reply src/flb_upstream.c]: DONE.

**Comment by erain** [comment]
Hi @edsiper , I've pushed a new commit which addressed all comments, and I also tested my new commit the same way as https://github.com/fluent/fluent-bit/pull/3272#issuecomment-805153418

File: include/fluent-bit/flb_config.h [hunk 4]
```diff
@@ -154,6 +154,16 @@ struct flb_config {
...
 
+    /*
+     * A comma-separated list of host names that shouldn't go through
+     * any proxy is set in (only an asterisk, * matches all hosts).
+     * As a convention (https://curl.se/docs/manual.html), this value can be set
+     * and respected by `NO_PROXY` environment variable when `HTTP_PROXY` is used.
+     * Example: NO_PROXY="127.0.0.1,localhost,kubernetes.default.svc"
+     * Note: only `,` is allowed as seperator between URLs.
+     */;
+    char *no_proxy;
```
Comment (frcai) [inline include/fluent-bit/flb_config.h:165]: Can we support IP ranges in CIDR format? Check https://docs.openshift.com/enterprise/3.2/install_config/http_proxies.html?

Comment (erain) [reply include/fluent-bit/flb_config.h:165]: I don't think this is feasible for now. This is because that we only have `hostname` when creating a `http_upstream`. I don't know a easy / feasible way for turning that into an IP address and do the CIDR check. 

(Golang supports that natively. Not so easy with the fluent-bit's C code base. )

If you have some ideas, feel free to enhance this feature based on my PR.

**Comment by edsiper** [comment]
hi @erain, after code review and CI fixes, is this ready to be merged from a functionality perspective ?

**Comment by erain** [comment]
> hi @erain, after code review and CI fixes, is this ready to be merged from a functionality perspective ?

Yes it is. 

Sorry that I was occupied in the past few days and haven't looked into the CI failed earlier. It looks like related to my attempt on enhancing the current unit test, and I believe I followed the existing pattern. Could you help me to see what did I miss?

File: include/fluent-bit/flb_config.h (hunk 4 above)
Comment (frcai) [reply include/fluent-bit/flb_config.h:165]: I think if IP address was configured in config file, the value of hostname should be IP address which you check against with CIDR.

Comment (erain) [reply include/fluent-bit/flb_config.h:165]: The hostname is something uncertain, depends on the remote host. For example, for stackdriver output plugin it will be the `logging.googleapis.com`. Almost certainly we will have a domain name instead of a IP address here, and since the whole fluent-bit's http client is implemented in its own codebase, doing DNS lookup etc. here would be non-trivial. 

So again, though this is nice to have, I would consider this as a separate feature requirement. Feel free to contribute PR for this.

**Comment by edsiper** [comment]
- This one is missing the signature on commits (DCO error) 
- Looks like there is a header broken ? check this: https://ci.appveyor.com/project/fluent/fluent-bit-2e87g/builds/38503081/job/aa7rivpn34uor77w#L1380
- unit test are broken: https://github.com/fluent/fluent-bit/pull/3272/checks?check_run_id=2298240264#step:5:1431

**Comment by erain** [comment]
Thanks @edsiper , I will do another commit ASAP (currently busy with other on-fire issues on my side)

File: include/fluent-bit/flb_config.h [hunk 5]
```diff
@@ -154,6 +154,16 @@ struct flb_config {
...
     char *http_proxy;
 
+    /*
+     * A comma-separated list of host names that shouldn't go through
+     * any proxy is set in (only an asterisk, * matches all hosts).
+     * As a convention (https://curl.se/docs/manual.html), this value can be set
+     * and respected by `NO_PROXY` environment variable when `HTTP_PROXY` is used.
+     * Example: NO_PROXY="127.0.0.1,localhost,kubernetes.default.svc"
+     * Note: only `,` is allowed as seperator between URLs.
+     */;
```
Comment (JeffLuoo) [inline include/fluent-bit/flb_config.h:164]: Hi Yu @erain Is this `;` a typo?

**Comment by JeffLuoo** [comment]
Hi Yu @erain Could you please remove the `;` as https://github.com/fluent/fluent-bit/pull/3272#discussion_r620826008 and re-commit to trigger the test again? 

Also don't forget the signoff `-s` option when committing : D (details: https://github.com/fluent/fluent-bit/pull/3272/checks?check_run_id=2241296333)

File: include/fluent-bit/flb_config.h (hunk 5 above)
Comment (erain) [reply include/fluent-bit/flb_config.h]: FIxed. PTAL.

**Comment by erain** [comment]
> Hi Yu @erain Could you please remove the `;` as [#3272 (comment)](https://github.com/fluent/fluent-bit/pull/3272#discussion_r620826008) and re-commit to trigger the test again?
> 
> Also don't forget the signoff `-s` option when commit : D (details: https://github.com/fluent/fluent-bit/pull/3272/checks?check_run_id=2241296333)

Many thanks!!

I fixed changes and now CI should pass. And I will prepare a doc PR

**Comment by JeffLuoo** [comment]
Ping @edsiper  Could you please take a look at this PR? Thanks

**Comment by qingling128** [comment]
@agup006 @edsiper Any idea about the failed integration test https://github.com/fluent/fluent-bit/pull/3272/checks?check_run_id=2462035748

**Comment by qingling128** [comment]
Re-triggered the failed test. Running now.

**Comment by qingling128** [comment]
`Run integration tests for PR / publish the docker images for PR (pull_request_target)` seems to have failed again: https://github.com/fluent/fluent-bit/pull/3272/checks?check_run_id=2521701599. It successfully built the container, but failed to push it to Dockerhub. That seems like a test infra issue though, probably not triggered by this PR.

**Comment by qingling128** [comment]
You are welcome! Do we have any planned release coming that would include this feature?

**Comment by edsiper** [comment]
we can do something early next week.

**Comment by qingling128** [comment]
Sounds great. Thanks @edsiper @erain @JeffLuoo for making this happen!
